File: src/store.ts
import { nanoid } from "nanoid";
import { BOARD_SIZE } from "./config";
import { Board, Room, SymbolXO, Presence } from "./types";
import { db } from "./db";
import { histories } from "./db/schema";
import { eq, sql } from "drizzle-orm";

export const rooms = new Map<string, Room>();
// export const histories = new Map<PairKey, Versus>(); // -> Đã bỏ, dùng DB

export function newBoard(): Board {
  return Array.from({ length: BOARD_SIZE }, () =>
    Array(BOARD_SIZE).fill(0 as 0)
  );
}

export function makeRoom(id?: string): Room {
  return {
    id: id || nanoid(6).toUpperCase(),
    board: newBoard(),
    players: [],
    nextTurn: "X",
    winner: null,
    members: new Map<string, Presence>(),
    offlineTimers: new Map(),
    firstMoverId: null,
    lastWinnerId: null,
  };
}

export function symbolOf(room: Room, userId: string): SymbolXO | null {
  const p = room.players.find((p) => p.id === userId);
  return p ? p.symbol : null;
}

export function pairKey(id1: string, id2: string): string {
  return [id1, id2].sort().join("|");
}

// Chuyển sang async để gọi DB
export async function incHistory(
  winnerId: string | null,
  id1: string,
  id2: string
) {
  const [A, B] = [id1, id2].sort();
  if (A === B) return; // không ghi lịch sử tự đấu với chính mình

  const pKey = `${A}|${B}`;

  // Chuẩn bị data update
  let updateData: any = {};
  if (winnerId === null) {
    updateData = { draws: sql`draws + 1` };
  } else if (winnerId === A) {
    updateData = { winsA: sql`wins_a + 1` };
  } else {
    updateData = { winsB: sql`wins_b + 1` };
  }

  // Upsert (Insert nếu chưa có, Update nếu có rồi)
  await db
    .insert(histories)
    .values({
      pairKey: pKey,
      playerA: A,
      playerB: B,
      winsA: winnerId === A ? 1 : 0,
      winsB: winnerId === B ? 1 : 0,
      draws: winnerId === null ? 1 : 0,
    })
    .onConflictDoUpdate({
      target: histories.pairKey,
      set: updateData,
    });
}

const historyQuery = db
  .select()
  .from(histories)
  .where(eq(histories.pairKey, sql.placeholder("pKey")))
  .prepare();

export function perspectiveHistory(me: string, other: string) {
  const pKey = pairKey(me, other);
  const h = historyQuery.get({ pKey }) || {
    playerA: me < other ? me : other,
    playerB: me < other ? other : me,
    winsA: 0,
    winsB: 0,
    draws: 0,
  };

  const meIsA = h.playerA === me;
  return {
    me,
    opponent: other,
    wins: meIsA ? h.winsA : h.winsB,
    losses: meIsA ? h.winsB : h.winsA,
    draws: h.draws,
    total: h.winsA + h.winsB + h.draws,
  };
}

export function pickNewHost(room: Room): string | undefined {
  // 1) ưu tiên player còn online
  for (const p of room.players) {
    const prez = room.members.get(p.id);
    if (prez && prez.sockets.size > 0) return p.id;
  }
  // 2) khán giả còn online
  for (const [uid, prez] of room.members) {
    if (prez.sockets.size > 0) return uid;
  }
  return undefined;
}

export function checkWinner(
  board: Board,
  x: number,
  y: number
): SymbolXO | null {
  const val = board[y][x]; // 1 or -1
  if (!val) return null;
  const dirs: Array<[number, number]> = [
    [1, 0],
    [0, 1],
    [1, 1],
    [1, -1],
  ];
  for (const [dx, dy] of dirs) {
    let count = 1,
      i = 1;
    while (true) {
      const nx = x + dx * i,
        ny = y + dy * i;
      if (nx < 0 || ny < 0 || nx >= BOARD_SIZE || ny >= BOARD_SIZE) break;
      if (board[ny][nx] !== val) break;
      count++;
      i++;
    }
    i = 1;
    while (true) {
      const nx = x - dx * i,
        ny = y - dy * i;
      if (nx < 0 || ny < 0 || nx >= BOARD_SIZE || ny >= BOARD_SIZE) break;
      if (board[ny][nx] !== val) break;
      count++;
      i++;
    }
    if (count >= 5) return val === 1 ? "X" : "O";
  }
  return null;
}

export function isBoardFull(board: Board) {
  for (let y = 0; y < BOARD_SIZE; y++)
    for (let x = 0; x < BOARD_SIZE; x++) if (board[y][x] === 0) return false;
  return true;
}

File: rust/src/store.rs
use crate::config::board_size;
use crate::db::Database;
use crate::types::*;
use parking_lot::RwLock;
use std::collections::{HashMap, HashSet};

/// Shared application state (mirrors rooms Map + store functions)
pub struct AppState {
    pub rooms: RwLock<HashMap<String, Room>>,
    pub db: Database,
}

impl AppState {
    pub fn new(db: Database) -> Self {
        AppState {
            rooms: RwLock::new(HashMap::new()),
            db,
        }
    }
}

/// Create a new empty board (mirrors newBoard)
pub fn new_board() -> Board {
    let size = board_size();
    vec![vec![0i8; size]; size]
}

/// Create a new room (mirrors makeRoom)
pub fn make_room(id: Option<String>) -> Room {
    let room_id = id.unwrap_or_else(|| {
        nanoid::nanoid!(6, &nanoid::alphabet::SAFE).to_uppercase()
    });

    Room {
        id: room_id,
        board: new_board(),
        players: Vec::new(),
        next_turn: SymbolXO::X,
        winner: None,
        last_move: None,
        host_id: None,
        members: HashMap::new(),
        offline_timers: HashSet::new(),
        first_mover_id: None,
        last_winner_id: None,
    }
}

/// Get symbol of a player in the room (mirrors symbolOf)
pub fn symbol_of(room: &Room, user_id: &str) -> Option<SymbolXO> {
    room.players.iter()
        .find(|p| p.id == user_id)
        .map(|p| p.symbol)
}

/// Generate pair key from two user IDs (mirrors pairKey)
pub fn pair_key(id1: &str, id2: &str) -> String {
    let mut ids = [id1, id2];
    ids.sort();
    format!("{}|{}", ids[0], ids[1])
}

/// Pick a new host when current host disconnects (mirrors pickNewHost)
pub fn pick_new_host(room: &Room) -> Option<String> {
    // 1) ưu tiên player còn online
    for p in &room.players {
        if let Some(prez) = room.members.get(&p.id) {
            if !prez.sockets.is_empty() {
                return Some(p.id.clone());
            }
        }
    }
    // 2) khán giả còn online
    for (uid, prez) in &room.members {
        if !prez.sockets.is_empty() {
            return Some(uid.clone());
        }
    }
    None
}

/// Check if there's a winner after a move (mirrors checkWinner)
pub fn check_winner(board: &Board, x: usize, y: usize) -> Option<SymbolXO> {
    let size = board_size();
    let val = board[y][x];
    if val == 0 {
        return None;
    }

    let dirs: [(i32, i32); 4] = [(1, 0), (0, 1), (1, 1), (1, -1)];

    for (dx, dy) in &dirs {
        let mut count = 1u32;

        // Forward direction
        let mut i: i32 = 1;
        loop {
            let nx = x as i32 + dx * i;
            let ny = y as i32 + dy * i;
            if nx < 0 || ny < 0 || nx >= size as i32 || ny >= size as i32 {
                break;
            }
            if board[ny as usize][nx as usize] != val {
                break;
            }
            count += 1;
            i += 1;
        }

        // Backward direction
        i = 1;
        loop {
            let nx = x as i32 - dx * i;
            let ny = y as i32 - dy * i;
            if nx < 0 || ny < 0 || nx >= size as i32 || ny >= size as i32 {
                break;
            }
            if board[ny as usize][nx as usize] != val {
                break;
            }
            count += 1;
            i += 1;
        }

        if count >= 5 {
            return if val == 1 {
                Some(SymbolXO::X)
            } else {
                Some(SymbolXO::O)
            };
        }
    }

    None
}

/// Check if the board is full (mirrors isBoardFull)
pub fn is_board_full(board: &Board) -> bool {
    let size = board_size();
    for y in 0..size {
        for x in 0..size {
            if board[y][x] == 0 {
                return false;
            }
        }
    }
    true
}
